app/src/main/java/org/thoughtcrime/securesms/messages/GroupSendUtil.java

```diff
@@ -0,0 +1,621 @@
+                                                                    ContentHint contentHint,
+                                                                    @NonNull SignalServiceDataMessage message)
+      throws IOException, UntrustedIdentityException
+  {
+    return sendMessage(context, groupId, null, allTargets, isRecipientUpdate, DataSendOperation.unresendable(message, contentHint), null);
+  }
+
+  /**
+   * Handles all of the logic of sending to a group. Will do sender key sends and legacy 1:1 sends as-needed, and give you back a list of
+   * {@link SendMessageResult}s just like we're used to.
+   *
+   * @param groupId The groupId of the group you're sending to, or null if you're sending to a collection of recipients not joined by a group.
+   */
+  @WorkerThread
+  public static List<SendMessageResult> sendTypingMessage(@NonNull Context context,
+                                                          @Nullable GroupId.V2 groupId,
+                                                          @NonNull List<Recipient> allTargets,
+                                                          @NonNull SignalServiceTypingMessage message,
+                                                          @Nullable CancelationSignal cancelationSignal)
+      throws IOException, UntrustedIdentityException
+  {
+    return sendMessage(context, groupId, null, allTargets, false, new TypingSendOperation(message), cancelationSignal);
+  }
+
+  /**
+   * Handles all of the logic of sending to a group. Will do sender key sends and legacy 1:1 sends as-needed, and give you back a list of
+   * {@link SendMessageResult}s just like we're used to.
+   *
+   * @param groupId The groupId of the group you're sending to, or null if you're sending to a collection of recipients not joined by a group.
+   */
+  @WorkerThread
+  public static List<SendMessageResult> sendCallMessage(@NonNull Context context,
+                                                        @Nullable GroupId.V2 groupId,
+                                                        @NonNull List<Recipient> allTargets,
+                                                        @NonNull SignalServiceCallMessage message)
+      throws IOException, UntrustedIdentityException
+  {
+    return sendMessage(context, groupId, null, allTargets, false, new CallSendOperation(message), null);
+  }
+
+  /**
+   * Handles all of the logic of sending to a group. Will do sender key sends and legacy 1:1 sends as-needed, and give you back a list of
+   * {@link SendMessageResult}s just like we're used to.
+   *
+   * @param groupId The groupId of the group you're sending to, or null if you're sending to a collection of recipients not joined by a group.
+   * @param isRecipientUpdate True if you've already sent this message to some recipients in the past, otherwise false.
+   */
+  @WorkerThread
+  private static List<SendMessageResult> sendMessage(@NonNull Context context,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 44. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 161-318 into a separate method.